test/extensions/filters/http/ext_proc/filter_test.cc

```diff
@@ -2261,6 +2272,90 @@ TEST_F(HttpFilterTest, FailOnInvalidHeaderMutations) {
+  initialize(R"EOF(
+  grpc_service:
+    envoy_grpc:
+      cluster_name: "ext_proc_server"
+  )EOF");
+  HttpConnectionManagerImplMixin::setup(false, "fake-server");
+  HttpConnectionManagerImplMixin::initial_buffer_limit_ = 10;
+  HttpConnectionManagerImplMixin::setUpBufferLimits();
+
+  std::shared_ptr<Http::MockStreamDecoderFilter> mock_filter(
+      new NiceMock<Http::MockStreamDecoderFilter>());
+  EXPECT_CALL(filter_factory_, createFilterChain(_))
+      .WillOnce(Invoke([&](Http::FilterChainManager& manager) -> bool {
+        // Add ext_proc filter.
+        Http::FilterFactoryCb cb = [&](Http::FilterChainFactoryCallbacks& callbacks) {
+          callbacks.addStreamDecoderFilter(filter_);
+        };
+        manager.applyFilterFactoryCb({}, cb);
+        // Add the mock-decoder filter.
+        Http::FilterFactoryCb mock_filter_cb = [&](Http::FilterChainFactoryCallbacks& callbacks) {
+          callbacks.addStreamDecoderFilter(mock_filter);
+        };
+        manager.applyFilterFactoryCb({}, mock_filter_cb);
+
+        return true;
+      }));
+  EXPECT_CALL(*mock_filter, decodeHeaders(_, false))
+      .WillOnce(Invoke([&](Http::RequestHeaderMap& headers, bool end_stream) {
+        // The next decoder filter should be able to see the mutations made by the external server.
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Path()->value().getStringView(), "/mutated_path/bluh");
+        EXPECT_EQ(headers.get(Envoy::Http::LowerCaseString("foo"))[0]->value().getStringView(),
+                  "gift-from-external-server");
+        mock_filter->callbacks_->sendLocalReply(Http::Code::OK, "Direct response from mock filter.",
+                                                nullptr, absl::nullopt, "");
+        return Http::FilterHeadersStatus::StopIteration;
+      }));
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, _))
+      .WillOnce(Invoke([&](const Http::ResponseHeaderMap& headers, bool end_stream) -> void {
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Status()->value().getStringView(), "200");
+      }));
+  EXPECT_CALL(response_encoder_, encodeData(_, _))
+      .WillOnce(Invoke([&](Buffer::Instance& data, bool end_stream) -> void {
+        EXPECT_TRUE(end_stream);
+        EXPECT_EQ(data.toString(), "Direct response from mock filter.");
+      }));
+  // Start the request.
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> Http::Status {
+    EXPECT_EQ(data.length(), 5);
+    data.drain(5);
+
+    decoder_ = &conn_manager_->newStream(response_encoder_);
+    Http::RequestHeaderMapPtr headers{new Http::TestRequestHeaderMapImpl{
+        {":authority", "host"}, {":path", "/bluh"}, {":method", "GET"}}};
+    decoder_->decodeHeaders(std::move(headers), false);
+    Buffer::OwnedImpl resp_body("Definitely more than 10 bytes data.");
+    decoder_->decodeData(resp_body, true);
```

Comment:
I thought that `decoder_` (which is a `RequestDecoder`) is handling client -> server request data, but here it is being fed `resp_body`. Should `resp_body` actually be named `req_body`, or am I confused?

Reply:
the decode_ here is inherited from the mixin, it's actually a "RequestDecoder", which is based by an ActiveStream



  ActiveStreamPtr new_stream(new ActiveStream(*this, response_encoder.getStream().bufferLimit(),
                                              std::move(downstream_stream_account)));

Reply:
But the [doc for `RequestEncoder`](https://github.com/envoyproxy/envoy/blob/6424fa91b82d4421d2d9dab48b60d831a61ee5d1/envoy/http/codec.h#L221) says that it is "used for receiving a request (client to server)", and here we are calling `decodeData` on it, which (given the type and doc) sounds like data in the client to server direction, i.e. the request body.

Reply:
when Envoy receives a http request with body, as a request decoder, it will need to decode the "headermap" first, then the "body" by using decodeData(). 

We are calling on decodeData() here directly, to mimick that the underneath codec has received the body, and triggers the body call.

Reply:
requestDecoder is also a streamDecoder, which defines the StreamDecoder::decodeData interface.

Reply:
I think we might be talking past one another:

> when Envoy receives a http request with body, as a request decoder, it will need to decode the "headermap" first, then the "body" by using decodeData().

Right, and is the body that is passed to the request decoder the request body or the response body? My understanding is that `RequestDecoder::decodeData` receives the request body. Is that correct?

If `RequestDecoder::decodeData` indeed receives the request body, then ISTM the variable here should be named `req_body`, not `resp_body`. That's the only point I was raising.

Reply:
my bad. 
need another coffee. 
updated. Thanks!